File: Problem74_N-Queens.java
// Time Complexity : O(n!)
// Space Complexity : O(n^2)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach
class Solution {
    List<List<String>> result;
    boolean[][] board;
    public List<List<String>> solveNQueens(int n) {
        result = new ArrayList<>();
        this.board = new boolean[n][n];
        backtrack(0);
        return result;
    }
    private void backtrack(int r) {
        // base case
        if(r == board.length) {
            List<String> li = new ArrayList<>();
            for(int i = 0; i < board.length; i++) {
                StringBuilder sb = new StringBuilder();
                for(int j = 0; j < board.length; j++) {
                    if(board[i][j] == true) {
                        sb.append('Q');
                    } else {
                        sb.append('.');
                    }
                }
                li.add(sb.toString());
            }
            result.add(li);
        }
        
        // logic
        for(int j = 0; j < board.length; j++) {
            if(isSafe(r, j)) {
                // action
                board[r][j] = true;
                // recurse 
                backtrack(r + 1);
                // backtrack
                board[r][j] = false;
            }
        }
    }
    
    private boolean isSafe(int row, int col) {
        // column check
        for(int i = 0; i < row; i++) {
            if(board[i][col] == true)
                return false;
        }
        
        // diagonal up left
        int i = row;
        int j = col;
        while(i >= 0 && j >= 0) {
            if(board[i][j] == true)
                return false;
            i--;
            j--;
        }
        
        // diagonal up right
        i = row;
        j = col;
        while(i >= 0 && j < board.length) {
            if(board[i][j] == true)
                return false;
            i--;
            j++;
        }
        return true;
    }
}

File: Problem-74 N-Queens.py
# 51. N-Queens
# https://leetcode.com/problems/n-queens/

# Time Complexiety: O(n!)
# Space Complexiety: O(n)

class Solution:
    def __init__(self):
        self.res = []

    def isSafe(self, board, row, col, n):
        # We will not check the row as in each row only we will put one queen so it cannot have more queens.
        # We will not check below our current row and col because that part is yet to have a queen.

        # Checking column
        for i in range(row):
            if board[i][col] == True:
                return False
        
        # Major Diagonal
        i = row
        j = col
        while i >= 0 and j >= 0:
            if board[i][j] == True:
                return False
            i -= 1
            j -= 1

        # Minor Diagonal
        i = row
        j = col
        while i >= 0 and j < n:
            if board[i][j] == True:
                return False
            i -= 1
            j += 1

        return True

    def _helper(self, row, n,board):
        # Base
        if row == n:
            # Format result
            li = list()
            for i in range(n):
                s = []
                for j in range(n):
                    if board[i][j] == True:
                        s.append("Q")
                    else:
                        s.append(".")
                li.append("".join(s))
            self.res.append(li.copy())
            return
        
        for col in range(n):
            if self.isSafe(board, row, col, n):
                board[row][col] = True
                self._helper(row+1, n, board)
                board[row][col] = False

    def solveNQueens(self, n):
        board = [[False for i in range(n)] for i in range(n)]

        self._helper(0, n, board)
        return self.res
